### DISCUSSION TITLE
uWebSocket Crashed when I enable websocket compression.

### DISCUSSION THREAD

**Discussion by balloonwj** [uNetworking/uWebSockets#930] [Q&A] [answered]
hi, I met the issue that uWebSocket Crashed when I enable websocket compression. Not always, but occassionally. I hava already checked my string to compress when crashed. The string to compress is  no problem.
My Code like below:
bool WebSocketSession::send(const char *pData, int dataLength, bool closeSession)
{   
        //The last param who equals to true indicates we enable websocket compression
        m_pwsHandle->send(std::string_view(pData, dataLength), uWS::TEXT, true);
        return true;
}
The crash stack is:
 
![image](https://user-images.githubusercontent.com/16314187/63759087-82e07d80-c8ef-11e9-8b6a-c814c60e9478.png)

Please help me, thanks very much.

**Comment by ghost** [comment, marked as answer]
Hello,

If you look at the deepest part of your stack; start_thread and follow the stack up to where it crashes in deflate_fast you see that there is no call to us_loop_run. In other words, you are not sending from the thread that owns the socket.

This library is strictly single threaded and needs one event-loop per thread, just like Node.js and similar. If you want to send from another thread you have to use us_wakeup_loop and send from the wakeup callback. However this incurs a few kernel calls and you also miss the fast path, since you need to buffer up the data somewhere until the wakeup callback can consume it.

**Comment by balloonwj** [comment]
Thank you for your reply.
 I know the principle of one event-loop per thread. But in your uWebSocket library how I can transfer data to the socket where owns the socket ? 
My situation is that my data have two sources. One of them are from the third system(not in threads which own sockets), and the other are produced by clients' request(in threads which own sockets).
Both data will push to websocket clients by websocket servers. I mean a WebSocket<bool, bool> class supply a send() interface, why does not the function of send() interface  put data in libuv event-loop?

**Comment by ghost** [comment]
Loop::defer(lambda) will call us_wakeup_loop and run that lambda on the thread which Loop is running on. That's the only place where you can send stuff.

```
Loop::defer([whatever_data]() {
// do your sending here
});
```

You can get the Loop of a socket by us_socket_context, then us_socket_context_loop then convert that us_loop_t * to Loop * and defer.

WebSocket<bool, bool> ptr is convertible to us_socket_t ptr. Then you get the us_loop_t ptr and convert that to a Loop ptr.

**Comment by balloonwj** [comment]
Thank you very much again.
According to what you said, did it means:
```
Loop::defer([whatever_data]() {
// do your sending here
//I use a WebSocket<bool, bool> object or ptr here to send data and set the last param to true 
//if I want to enable websocket compression? Is the right code below?
bool compressFlag = true;
//m_pwsHandle is a WebSocket<bool, bool> object 
m_pwsHandle->send(mydata, uWS::TEXT, compressFlag);
});
```

**Comment by aFoxWithDream** [comment]
hello, i has the same problem, i used another thread and user the:
uWS::Loop::get()->defer([this, message]() {
        if (!m_pClient)
        {
            return;
        }
        //this->m_pClient is a WebSocket<bool, bool> object
        this->m_pClient->send(message, uWS::OpCode::BINARY);
    });

but the client cannot reveive msg,  how can i solve this question.
